Add shallow cloning capability

blyxyas opened the pull request:
Add shallow cloning (as well as customizable depth), introduced in `libgit2 v1.7.0`

ehuss commented on src/build.rs:
```diff
@@ -171,6 +171,28 @@ impl<'cb> RepoBuilder<'cb> {
         self
     }
 
+    /// Specify fetch depth, a value less than zero is interpreted as pull everything (effectively the same as no declaring a limit depth)
+    pub fn depth(&mut self, depth: i32) -> &mut RepoBuilder<'cb> {
```
Instead of adding a new method here, would it be possible to leave this as a setting that required calling `.fetch_options(...)` instead? I am reluctant to try to repeat low-level option settings, since that would require deciding which settings are or are not replicated, and possibly causing confusion when something could be specified in multiple places.

blyxyas replied:
So this should be deleted? I find the `depth` parameter to be kinda (very) important imo, but it can be removed without any problems.

ehuss commented on src/build.rs:
```diff
@@ -171,6 +171,28 @@ impl<'cb> RepoBuilder<'cb> {
+        }
+
+        self
+        // if let Some(ref fetch_opts) = &self.fetch_opts {
+        //     fetch_opts.depth(depth);
+        // } else {
+        //     let mut fo = FetchOptions::new();
+        //     fo.depth(depth);
+        //     self.fetch_options(fo);
+        // }
+        // self
```
Looks to be some vestigial work?

ehuss commented on src/repo.rs:
```diff
@@ -341,6 +341,13 @@ impl Repository {
         Ok(repo)
     }
 
+    /// Clone a remote repository with a depth of one (shallow cloning)
+    ///
+    /// Similar to `git clone --depth 1`
+    pub fn shallow_clone<P: AsRef<Path>>(url: &str, into: P) -> Result<Repository, Error> {
```
Similar to the comment on `RepoBuilder`, I think I would like to avoid having too many convenience functions, since there is a somewhat combinatorial explosion of all the different options someone might want. Using a builder pattern can avoid adding all sorts of variants of different methods.

ehuss reviewed (commented):
Thanks for adding this!

Can you also include a test for it?

blyxyas commented:
I'm not sure how to add a test for this, it seems like there aren't any tests for cloning features.

ehuss commented:
> I'm not sure how to add a test for this, it seems like there aren't any tests for cloning features.

At the bottom of `remote.rs` there are various tests which set `RemoteCallbacks` options. I would set up a test that creates a repository with 2 commits. Then clone it with depth=1, and verify there is only one commit in the clone. The `test.rs` module contains some testing utilities for creating commits and such.

blyxyas commented:
This new commit addresses the comments and contains a first attempt to test it. It should work but it doesn't, I need some guidance here.

The test creates a repo and commits two times to that new repo. Then creates a `TempDir` to clone the original repo using fetch options with `depth` set to 1. But when checking for the first commit that shouldn't exist because `depth` is 1, it's still there.

rawhuul commented:
Any work on this?

ehuss commented:
Unfortunately it looks like libgit2 does not do shallow clones on local repositories. I think it requires the smart protocol and relies on the server to figure out what to fetch. I don't think we have any testing infrastructure for the smart protocol, so I think we'll just need to not include a test for this for now.

I have filed https://github.com/libgit2/libgit2/issues/6634 on that issue.

blyxyas commented:
So for now, it can go without a test? (I'll add a fixme)

ehuss commented:
Yep.

rawhuul commented:
Why it has not been merged yet?

ehuss reviewed (approved):
Thanks!
